src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Debugging/DefaultRazorBreakpointResolver.cs

```diff
@@ -0,0 +1,144 @@
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Debugging
+{
+    [Export(typeof(RazorBreakpointResolver))]
+    internal class DefaultRazorBreakpointResolver : RazorBreakpointResolver
+    {
+        private readonly FileUriProvider _fileUriProvider;
+        private readonly LSPDocumentManager _documentManager;
+        private readonly LSPProjectionProvider _projectionProvider;
+        private readonly LSPDocumentMappingProvider _documentMappingProvider;
+        private readonly CSharpBreakpointResolver _csharpBreakpointResolver;
+
+        [ImportingConstructor]
+        public DefaultRazorBreakpointResolver(
+            FileUriProvider fileUriProvider,
+            LSPDocumentManager documentManager,
+            LSPProjectionProvider projectionProvider,
+            LSPDocumentMappingProvider documentMappingProvider,
+            CSharpBreakpointResolver csharpBreakpointResolver)
+        {
+            if (fileUriProvider is null)
+            {
+                throw new ArgumentNullException(nameof(fileUriProvider));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (projectionProvider is null)
+            {
+                throw new ArgumentNullException(nameof(projectionProvider));
+            }
+
+            if (documentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(documentMappingProvider));
+            }
+
+            if (csharpBreakpointResolver is null)
+            {
+                throw new ArgumentNullException(nameof(csharpBreakpointResolver));
+            }
+
+            _fileUriProvider = fileUriProvider;
+            _documentManager = documentManager;
+            _projectionProvider = projectionProvider;
+            _documentMappingProvider = documentMappingProvider;
+            _csharpBreakpointResolver = csharpBreakpointResolver;
+        }
+
+        public override async Task<Range> TryResolveBreakpointRangeAsync(ITextBuffer textBuffer, int lineIndex, int characterIndex, CancellationToken cancellationToken)
+        {
+            if (textBuffer is null)
+            {
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            if (!_fileUriProvider.TryGet(textBuffer, out var documentUri))
+            {
+                // Not an addressable Razor document. Do not allow a breakpoint here.
```

Comment:
Could we get examples of what a reasonable way each of these `return null;` cases could happen in the comments? Maybe I'm just being thick but I don't really follow how we could receive a request to break on razor for a file that isn't even in our provider.

Reply:
Honestly these are mostly just being defensive.